### PR TITLE
Entity-Change: Interactions Armor-stands, Item frams | Dye mobs Sheep, Cats, Dogs

### PR DESCRIPTION
**Pull request by Altirix**
- Item Frames - Interactions logged, Rotation, Item Insert, Item Removal, Block Squash, Block Attach Gone. Entity Kill

- Armor Stands - Interactions logged, Item Insert, Item Removal, Entity Kill. (Tried to do Movement Logs but... that's not happening)

- Sheep - Interactions logged, Dye, Evoker Dye, Shear

- Dogs/Cats - Interactions logged, Dye

- Snow Golem - Interactions logged, Shear (Pumpkin)

- Entity-Change action type

ExtraData currently stores the ObjectIdentifier for the item used/removed

(Old)ObjectIdentifier stores entity id

all entity mixins are under the entities folder like blocks

entityState & oldEntityState store all entity tags before and after the change

### CODE REVIEW

File: src/main/kotlin/com/github/quiltservertools/ledger/actions/EntityModifyActionType.kt [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+import net.minecraft.item.Items
+import net.minecraft.server.MinecraftServer
+import net.minecraft.text.HoverEvent
+import net.minecraft.text.MutableText
+import net.minecraft.text.Text
+import net.minecraft.text.TranslatableText
+import net.minecraft.util.Util
+import net.minecraft.util.registry.Registry
+
+const val EQUIP = "equip"
```
Comment (yitzy299) [inline src/main/kotlin/com/github/quiltservertools/ledger/actions/EntityModifyActionType.kt:18]: Should be in `Sources`

File: src/main/kotlin/com/github/quiltservertools/ledger/utility/NbtUtils.kt [hunk 2]
```diff
@@ -21,4 +38,98 @@ object NbtUtils {
 
         return NbtHelper.toBlockState(stateTag)
     }
+
+    fun itemToProperties(item: ItemStack): NbtCompound? {
+        val itemTag = NbtCompound()
+
+        val storedNBT = item.writeNbt(NbtCompound())
+
+        Ledger.logger.info(storedNBT)
```
Comment (yitzy299) [inline src/main/kotlin/com/github/quiltservertools/ledger/utility/NbtUtils.kt:47]: Should be removed

File: src/main/kotlin/com/github/quiltservertools/ledger/actions/EntityModifyActionType.kt [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    override fun rollback(server: MinecraftServer): Boolean {
+        val world = server.getWorld(world)
+        val uuid = NbtUtils.entityFromProperties(extraData)!!.getUuid(UUID) ?: return false
+        val entity = world?.getEntity(uuid) ?: return false
+
+        val rollbackStack = NbtUtils.itemFromProperties(extraData, oldObjectIdentifier)
+
+        if (entity is ArmorStandEntity) {
+            // what items does this break with?
```
Comment (yitzy299) [inline src/main/kotlin/com/github/quiltservertools/ledger/actions/EntityModifyActionType.kt:105]: Should be removed

File: src/main/kotlin/com/github/quiltservertools/ledger/actions/EntityModifyActionType.kt [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+    override fun restore(server: MinecraftServer): Boolean {
+        val world = server.getWorld(world)
+        val uuid = NbtUtils.entityFromProperties(extraData)!!.getUuid(UUID) ?: return false
+        val entity = world?.getEntity(uuid) ?: return false
+
+        val rollbackStack = NbtUtils.itemFromProperties(extraData, oldObjectIdentifier)
+
+        if (entity is ArmorStandEntity) {
+            // what items does this break with?
```
Comment (yitzy299) [inline src/main/kotlin/com/github/quiltservertools/ledger/actions/EntityModifyActionType.kt:133]: Same as above

File: src/main/kotlin/com/github/quiltservertools/ledger/utility/NbtUtils.kt [hunk 5]
```diff
@@ -21,4 +38,98 @@ object NbtUtils {
...
+        if (tagNbt.contains(COUNT)) { itemTag.putByte(COUNT, tagNbt.getByte(COUNT)) }
+        else { itemTag.putByte(COUNT, 1) }
+
+        if (tagNbt.contains(TAG)) { itemTag.put(TAG, tagNbt.getCompound(TAG)) }
+
+        return ItemStack.fromNbt(itemTag)
+    }
+
+    //This monster will find non default NBT tags & log them
+    fun entityToProperties(entity: Entity, actionType: String): NbtCompound? {
```
Comment (yitzy299) [inline src/main/kotlin/com/github/quiltservertools/ledger/utility/NbtUtils.kt:76]: Why is this needed?

Any unused NBT tags (at least in vanilla) will do nothing if there's no functionality for them. There's no need to validate any NBT

File: src/main/java/com/github/quiltservertools/ledger/mixin/callbacks/LivingEntityMixin.java [hunk 6]
```diff
@@ -15,7 +16,7 @@
             at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/damage/DamageSource;getAttacker()Lnet/minecraft/entity/Entity;")
     )
     private void ledgerEntityKillInvoker(DamageSource source, CallbackInfo ci) {
-        LivingEntity entity = (LivingEntity) (Object) this;
+        Entity entity = (Entity) (Object) this;
```
Comment (yitzy299) [inline src/main/java/com/github/quiltservertools/ledger/mixin/callbacks/LivingEntityMixin.java:19]: Unnecessary change - LivingEntity extends Entity

**Review by yitzy299 — CHANGES_REQUESTED** [review]

**Review by yitzy299 — CHANGES_REQUESTED** [review]
I managed to not mention the following in my review:

- There's no logging for item frame and armour stand spawning
- The branch needs to be updated with our changes on master
- Every file needs to be run through a formatter, this formatting at the moment is not merge-ready

File: src/main/kotlin/com/github/quiltservertools/ledger/utility/NbtUtils.kt (hunk 5 above)
Comment (Altirix) [reply src/main/kotlin/com/github/quiltservertools/ledger/utility/NbtUtils.kt:76]: Originally the plan was to try to reduce how much was being stored to the db, armor stands mainly get quite large when fully equipped with enchanted armor and swapping items would normally mean the entire entity and the item get logged, when for those actions really all that is needed is the UUID and the item.

wanted to see if it could be applied to other parts that use extra data like items and entities that died but eh, it's still pretty rough around the edges for that, the checks are insane, I think I was just overly concerned if somehow a property for an entity got onto an item.

**Comment by RandomGgames** [comment]
Logging item frames (at least placement and breaking for now please!!)
